File: clients/ts/signalr/src/HubConnectionBuilder.ts
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

import { HttpConnection, IHttpConnectionOptions } from "./HttpConnection";
import { HubConnection, JsonHubProtocol } from "./HubConnection";
import { IHubProtocol } from "./IHubProtocol";
import { ILogger, LogLevel } from "./ILogger";
import { NullLogger } from "./Loggers";
import { ConsoleLogger } from "./Utils";

export class HubConnectionBuilder {
    private protocol: any;
    private httpConnectionOptions: IHttpConnectionOptions;
    private url: string;
    private logger: ILogger;

    public configureLogging(logging: LogLevel | ILogger): HubConnectionBuilder {
        if (isLogger(logging)) {
            this.logger = logging;
        } else {
            this.logger = new ConsoleLogger(logging);
        }

        return this;
    }

    public withUrl(url: string): HubConnectionBuilder;
    public withUrl(url: string, options: IHttpConnectionOptions): HubConnectionBuilder;
    public withUrl(url: string, options?: IHttpConnectionOptions): HubConnectionBuilder {
        this.url = url;
        this.httpConnectionOptions = options;
        return this;
    }

    public withHubProtocol(protocol: IHubProtocol): HubConnectionBuilder {
        this.protocol = protocol;
        return this;
    }

    public build(): HubConnection {
        // If httpConnectionOptions has a logger, use it. Otherwise, override it with the one
        // provided to configureLogger
        const httpConnectionOptions = this.httpConnectionOptions || {};

        // If it's 'null', the user **explicitly** asked for null, don't mess with it.
        if (httpConnectionOptions.logger === undefined) {
            // If our logger is undefined or null, that's OK, the HttpConnection constructor will handle it.
            httpConnectionOptions.logger = this.logger;
        }

        // Now create the connection
        if (!this.url) {
            throw new Error("The 'HubConnectionBuilder.withUrl' method must be called before building the connection.");
        }
        const connection = new HttpConnection(this.url, httpConnectionOptions);

        return new HubConnection(
            connection,
            this.logger || NullLogger.instance,
            this.protocol || new JsonHubProtocol());
    }
}

function isLogger(logger: any): logger is ILogger {
    return logger.log !== undefined;
}

File: clients/java/signalr/src/main/java/com/microsoft/signalr/HubConnectionBuilder.java
// Copyright (c) .NET Foundation. All rights reserved.
// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.

package com.microsoft.signalr;

import java.util.concurrent.CompletableFuture;
import java.util.function.Supplier;

public class HubConnectionBuilder {
    private String url;
    private Transport transport;
    private Logger logger;
    private HttpClient httpClient;
    private boolean skipNegotiate;
    private Supplier<CompletableFuture<String>> accessTokenProvider;

    private HubConnectionBuilder(String url) {
        this.url = url;
    }

    public static HubConnectionBuilder create(String url) {
        if (url == null || url.isEmpty()) {
            throw new IllegalArgumentException("A valid url is required.");
        }
        return new HubConnectionBuilder(url);
    }

    public HubConnectionBuilder withTransport(Transport transport) {
        this.transport = transport;
        return this;
    }


    public HubConnectionBuilder withHttpClient(HttpClient httpClient) {
        this.httpClient = httpClient;
        return this;
    }

    public HubConnectionBuilder configureLogging(LogLevel logLevel) {
        this.logger = new ConsoleLogger(logLevel);
        return this;
    }

    public HubConnectionBuilder withSkipNegotiate(boolean skipNegotiate){
        this.skipNegotiate = skipNegotiate;
        return this;
    }

    public HubConnectionBuilder withAccessTokenProvider(Supplier<CompletableFuture<String>> accessTokenProvider) {
        this.accessTokenProvider = accessTokenProvider;
        return this;
    }

    public HubConnectionBuilder configureLogging(Logger logger) {
        this.logger = logger;
        return this;
    }

    public HubConnectionBuilder withLogger(Logger logger) {
        this.logger = logger;
        return this;
    }

    public HubConnection build() {
        return new HubConnection(url, transport, skipNegotiate, logger, httpClient, accessTokenProvider);
    }
}